Creating a SSLContext per scope of routes

jitsni opened the pull request.

jitsni committed:
Using per route keystore/truststore

jitsni committed:
Merge remote-tracking branch 'upstream/develop' into store.per.route

jitsni committed:
Using the route's store while creating SSLContext

jitsni committed:
Removing SSLContext for a route when all corresponding routes are gone

jitsni committed:
Adding scopes to routes

jitsni committed:
Renaming store to scopeId changes

jitsni committed:
Adding tests for default scope and client auth

jitsni committed:
Merge remote-tracking branch 'upstream/develop' into store.per.route

Conflicts:
	pom.xml
	src/test/java/org/reaktivity/nukleus/tls/internal/streams/ClientIT.java

jitsni committed:
Minor renaming

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/TlsConfiguration.java:
```diff
@@ -28,10 +28,13 @@
     public static final String KEY_MANAGER_ALGORITHM_DEFAULT = "PKIX";
 
+    public final Configuration config;
+
```
Should be unnecessary, please remove.

jfallows replied:
Also `keyManagerAlgorithm()` below should be `public`.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/TlsNukleusFactorySpi.java:
```diff
@@ -63,18 +63,14 @@ public Nukleus create(
             new ServerStreamFactoryBuilder(
-                tlsConfig,
-                context);
+                tlsConfig);
```
Unwrap to previous line.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/TlsNukleusFactorySpi.java:
```diff
@@ -63,18 +63,14 @@ public Nukleus create(
             new ClientStreamFactoryBuilder(
-                tlsConfig,
-                context);
+                tlsConfig);
```
Unwrap to previous line.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/TlsNukleusFactorySpi.java:
```diff
@@ -83,17 +79,18 @@ public Nukleus create(
     }
 
-    private SSLContext initContext(
+    public static SSLContext initContext(
```
Why `public`?

jitsni replied:
Now, the SSLContext is created in factory builders and they are in different package. Hence made the method public.

One option is create an abstract factory builder and move the context creation there. That way both ClientStreamFactoryBuilder and ServerStreamFactoryBuilder can extend abstract factory builder and share the creation of SSLContext.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -317,6 +322,7 @@ private RouteFW wrapRoute(
         private long applicationTraceId;
 
         private ClientAcceptStream(
+            SSLEngine tlsEngine,
```
How about we pass in `Supplier<SSLEngine> supplyTlsEngine` and defer creation of `tlsEngine` until handling `BEGIN`?

jitsni replied:
Source#handleBegin() creates ClientAcceptStream and hands over BEGIN to newly created stream. Since BEGIN is following immediately, is there any benefit in delaying the creation of SSLEngine ?

jitsni replied:
Not creating Supplier<SSLEngine> as it was already a field, now becomes final field

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactoryBuilder.java:
```diff
@@ -147,9 +155,26 @@ public boolean handleRoute(int msgTypeId, DirectBuffer buffer, int index, int le
+                final TlsRouteExFW routeEx = route.extension().get(tlsRouteExRO::wrap);
+                final String scope = routeEx.scopeId().asString();
+                routes.merge(scope, 1, (old, inc) -> old + inc);
+                context.computeIfAbsent(scope, (x) -> initContext(config.config.directory(), config, scope));
```
`config.directory()` should work here instead, right?

jitsni replied:
Yes.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactoryBuilder.java:
```diff
@@ -33,15 +35,21 @@
     private final TlsConfiguration config;
-    private final SSLContext context;
+    private final Map<String, SSLContext> context;
+    private final Map<String, Integer> routes;
```
Use `MutableInteger` instead of `Integer`?

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactoryBuilder.java:
```diff
@@ -147,9 +155,26 @@ public boolean handleRoute(int msgTypeId, DirectBuffer buffer, int index, int le
+                final TlsRouteExFW routeEx = route.extension().get(tlsRouteExRO::wrap);
+                final String scope = routeEx.scopeId().asString();
+                routes.merge(scope, 1, (old, inc) -> old + inc);
+                context.computeIfAbsent(scope, (x) -> initContext(config.config.directory(), config, scope));
```
No parens on `(x)` in lambda.

jitsni replied:
removed the parens

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactoryBuilder.java:
```diff
@@ -147,9 +155,26 @@ public boolean handleRoute(int msgTypeId, DirectBuffer buffer, int index, int le
+                final TlsRouteExFW routeEx = unroute.extension().get(tlsRouteExRO::wrap);
+                final String scope = routeEx.scopeId().asString();
+                routes.merge(scope, -1, (old, inc) -> old + inc);
+                if (routes.get(scope) == 0)
```
Use return value of `merge` instead of repeating looking via `.get(scope)`.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactoryBuilder.java:
```diff
@@ -147,9 +155,26 @@ public boolean handleRoute(int msgTypeId, DirectBuffer buffer, int index, int le
                 final UnrouteFW unroute = unrouteRO.wrap(buffer, index, index + length);
+                final TlsRouteExFW routeEx = unroute.extension().get(tlsRouteExRO::wrap);
+                final String scope = routeEx.scopeId().asString();
+                routes.merge(scope, -1, (old, inc) -> old + inc);
```
Consider `routes.computeIfPresent(scope, (s, c) -> decrement(c));`, where:
```
private static MutableInteger decrement(
    MutableInteger count)
{
    count.value--;
    return count;
}
```

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -98,7 +99,7 @@
     private final ResetFW.Builder resetRW = new ResetFW.Builder();
 
-    private final SSLContext context;
+    private final Map<String, SSLContext> context;
```
Rename `context` to `contextsByScope`.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -121,7 +122,7 @@
     public ServerStreamFactory(
         TlsConfiguration config,
-        SSLContext context,
+        Map<String, SSLContext> context,
```
Rename `context` to `contextsByScope`.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactoryBuilder.java:
```diff
@@ -148,9 +155,26 @@ public boolean handleRoute(int msgTypeId, DirectBuffer buffer, int index, int le
                 final UnrouteFW unroute = unrouteRO.wrap(buffer, index, index + length);
+                final TlsRouteExFW routeEx = unroute.extension().get(tlsRouteExRO::wrap);
+                final String scope = routeEx.scopeId().asString();
+                routes.merge(scope, -1, (old, inc) -> old + inc);
```
See feedback on `ClientStreamFactoryBuilder`.

jitsni committed:
Minor misc changes:
- s/context/contextsByScope
- s/routes/routesByScope
- MutableInteger to keep track of routes count
- Using TlsConfig#directory() directly

jitsni committed:
directory need not be passed as parameter

jitsni committed:
Moving SSLContext creation to ServerStreamFactoryBuilder

jitsni committed:
s/scopeId/store in route extension

jitsni committed:
Making SSLEngine field final

jfallows committed:
Update nukleus-tls.spec dependency version

jitsni committed:
Moving timeout to outer rule

jitsni committed:
Moving timeout to outer rule

jitsni committed:
Adding client and server stores
